{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2ca9cc5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b54570d",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_bam = pysam.AlignmentFile(\"/tmp/Mazutislab-out/Ignas/RT_comparison/25_SSCV_KG_01_S1/25_SSCV_KG_01_S1_filtered.sorted.bam\", \"rb\")\n",
    "fasta = pysam.FastaFile(\"/tmp/Mazutislab-out/Ignas/RT_comparison/genome_and_annotations/Homo_sapiens.GRCh38.dna.primary_assembly.fa.modified\")\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "241ef1c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-\n",
      "sequence of read: AGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGTTGAG\n",
      "sequence of gene: AGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAG\n",
      "is secondary: False\n",
      "is reverese_strand: False\n",
      "248956422\n"
     ]
    }
   ],
   "source": [
    "for i,read in enumerate(input_bam.fetch()):\n",
    "    if i==1:\n",
    "        #find if the read was aligned to a particular gene:\n",
    "        print(read.get_tag('GN'))\n",
    "        print(f\"sequence of read: {read.query_sequence}\")\n",
    "        print(f\"sequence of gene: {fasta.fetch(read.reference_name, read.reference_start, read.reference_end)}\")\n",
    "        print(f\"is secondary: {read.is_secondary}\")\n",
    "        print(f\"is reverese_strand: {read.is_reverse}\")\n",
    "        print(f\"{fasta.get_reference_length(read.reference_name)}\")\n",
    "\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4ee66d2d",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "db = gffutils.create_db(\n",
    "    \"/tmp/Mazutislab-out/Ignas/RT_comparison/genome_and_annotations/gencode.v41.primary_assembly.annotation.gtf.filtered\",\n",
    "    dbfn=\"annotation.db\",\n",
    "    force=True,\n",
    "    keep_order=True,\n",
    "    disable_infer_transcripts=True,\n",
    "    disable_infer_genes=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ab4afdb7",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[<Feature exon (GRCh38_chr1:89551-90050[-]) at 0x7f6c1078edd0>, <Feature exon (GRCh38_chr1:90287-91105[-]) at 0x7f6c1078d1d0>]\n",
      "2\n",
      "89551\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "tid = \"GRCh38_ENST00000495576\"\n",
    "exons = list(db.children(tid, featuretype='exon', order_by='start'))\n",
    "print(exons)\n",
    "print(int(exons[0].attributes.get(\"exon_number\")[0]))\n",
    "print(exons[0].start)\n",
    "print(int(exons[1].attributes.get(\"exon_number\")[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "32753141",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11929/11929 [00:06<00:00, 1712.82it/s]\n"
     ]
    }
   ],
   "source": [
    "import tqdm\n",
    "#get transcript ids of lncRNAs with single isoform:\n",
    "lncs_with_single_isoform = {}\n",
    "for line in open('/tmp/Mazutislab-out/Ignas/RT_comparison/genome_and_annotations/lncs_with_single_isoform_gene_ids_names_transcripts.txt'):\n",
    "    gene_name_and_id = line.strip()\n",
    "    gene_id, gene_name, transcript_id = gene_name_and_id.split(\" \")\n",
    "    lncs_with_single_isoform[gene_name] = transcript_id\n",
    "transcripts = list(lncs_with_single_isoform.values())\n",
    "\n",
    "def get_exon_number(exon):\n",
    "    return int(exon.attributes.get(\"exon_number\")[0])\n",
    "\n",
    "#get fasta sequences of exonic regions, while taking strand into account\n",
    "def reverse_complement(seq):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}\n",
    "    return ''.join(complement[base] for base in reversed(seq))\n",
    "\n",
    "fasta_sequences = {}\n",
    "intron_positions = {}\n",
    "exon_positions = {}\n",
    "transcript_positions = {}\n",
    "for transcript_id in tqdm.tqdm(transcripts):\n",
    "#for transcript_id in [\"GRCh38_ENST00000495576\"]: #for tetsing\n",
    "    exons = list(db.children(transcript_id, featuretype='exon'))\n",
    "    transcript = db[transcript_id]\n",
    "    exons.sort(key=get_exon_number)\n",
    "    #print(exons)\n",
    "    exon_sequences = []\n",
    "    intron_positions[transcript_id] = []\n",
    "    exon_positions[transcript_id] = []\n",
    "    transcript_positions[transcript_id] = (transcript.start, transcript.end)\n",
    "    for i, exon in enumerate(exons):\n",
    "        #print(exon)\n",
    "        exon_positions[transcript_id].append((exon.start, exon.end))\n",
    "        seq = fasta.fetch(exon.chrom, exon.start - 1, exon.end)  # -1 because pysam is 0-based\n",
    "        if exon.strand == '-':\n",
    "            seq = reverse_complement(seq)\n",
    "        exon_sequences.append(seq)\n",
    "        # also for each sequence build a list of intron positions\n",
    "        if len(exon_sequences) > 1:\n",
    "            if exon.strand == '+':\n",
    "                prev_exon = exons[i - 1]\n",
    "                intron_start = prev_exon.end + 1\n",
    "                intron_end = exon.start - 1\n",
    "                intron_positions[transcript_id].append((intron_start, intron_end))\n",
    "            else:\n",
    "                prev_exon = exons[i - 1]\n",
    "                intron_start = exon.end + 1 \n",
    "                intron_end = prev_exon.start - 1\n",
    "                intron_positions[transcript_id].append((intron_start, intron_end))\n",
    "    full_sequence = ''.join(exon_sequences)\n",
    "    fasta_sequences[transcript_id] = full_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "ed0bcd9d",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(90051, 90286)]\n",
      "[(90287, 91105), (89551, 90050)]\n",
      "(89551, 91105)\n"
     ]
    }
   ],
   "source": [
    "len(fasta_sequences[tid])\n",
    "print(intron_positions[tid])\n",
    "print(exon_positions[tid])\n",
    "print(transcript_positions[tid])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "a0252c52",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "seq_2='TCAGCCTCCCAAGTAGCTGGGGCTACAGGCACCTGCCACCAAACCCGGCTAATTTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTAGCCAGGATCGTCTTGATCTCCTGACCTTGTGATCCACCCGCCTCGGCCTCCCAAATTGCTGGGATTACAGATGTGAGCCACCGCACCTGGTCCAAGAACCCAAGTTTTAGATCTAGAGTGATGTCAGCATGACATTGATTTCCTGAGGCCCAGGGGCGAAGGAGCTGAGGACAGCAGAGGGGTGAAGGAACTCAGCTACAGACAGCAGCAGCTGATGCACAGGCCTCCCAGCGCCTGAAGTCACCCGGAATTGGGAAGTGCTCAGAAGCTTACAAAGCTGCCTCGAGGTGGGAACATGACATAAATCCAAGAGCAGATCCCTGATCCTATAAAAATGTACTAGATGCAGTGGGGGCATTTTAAATGAGCAGAGAAGGACAGACAGATAAACAGAAGGACAAACAGTATTGGGATTGGGATAAATGCTCAGCTTTTGCCCAAATCTTAGTGACTTAAGCATCACTTATTTGCTCACGATTCTGTGGCTGGACCATTTGGTTTGGCTCACAGGGCAGGGACTGTGCTGGTCTTACCTGAGCAGACCTGCATGTCTGCGGTCAACTGGGTTGGCAGAGACAGAGTGACTGTCTTCCTCCAGGAAGCAGCAGGTTAACTGGTTGGCAGAGACAGAGGGACAGAGGGACTGTCTTCCTCCAGGAAGCAGCAGGTTAACTGGTTGGCAGAGACAGAGGGACAGAGGGACTGTCTTCCTCCAGGAAGCAGCAGGTTGGCTCTGTTTCCTTCGTGGGGCAGCTGGTCTCCAGGGCAGCAAGAGAGACCAAGCCCCAGTGCACATTCTACAGCCTCTGTGCACATCAGACTTGTTAATATCCCATTGGCCAGTGTAAGTCACTTGGCCAAGCCCAGATTAAGGAGTGGAAAGATGGAGGCTATCTCCTCCTGGGAGAGGAGGCAAAGGAGGTGGGAGTATTATGTGGCCACTTATGTTTGCAATCTACCATACTTAGCACTTTGAGAAAAGAATTAACTGAGAAACTTGCTTCAAATAGGGCCAGTAAAATGAAGCCCCAATTGAAGTAAAATGCATATATAAAAAATGAAACTGTGACCGATTTTAAGGACAGTATTGGCAAATATTTCTGTGCTCTTGGAGGAGAAGACCCTTATTGGCATGACATGTCAGAAACCACAATGAAAGAATTATTTTAACTTGCATTCATAAAAATTAAAATTATTCATTAAAAACATCGTGAATGAAATTAAAA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "bd93e441",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get number of mismatches between two strings\n",
    "def count_mismatches(seq1, seq2):\n",
    "    mismatches = 0\n",
    "    for a, b in zip(seq1, seq2):\n",
    "        if a != b:\n",
    "            mismatches += 1\n",
    "    return mismatches\n",
    "count_mismatches(seq_2, fasta_sequences[tid])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "d17a61ec",
   "metadata": {
    "vscode": {
     "languageId": "ruby"
    }
   },
   "outputs": [],
   "source": [
    "#write sequences and exon, intron and transcript regions to a file:\n",
    "with open('/tmp/Mazutislab-out/Ignas/RT_comparison/genome_and_annotations/lnc_fasta_sequences_exon_intron_positions.txt', 'w') as f:\n",
    "    for transcript_id in fasta_sequences:\n",
    "        f.write(f\">{transcript_id}\\n\")\n",
    "        f.write(f\"Sequence: {fasta_sequences[transcript_id]}\\n\")\n",
    "        f.write(f\"Transcript position: {transcript_positions[transcript_id]}\\n\")\n",
    "        f.write(f\"Exon positions: {exon_positions[transcript_id]}\\n\")\n",
    "        f.write(f\"Intron positions: {intron_positions[transcript_id]}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (star-env)",
   "language": "python",
   "name": "star-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
