{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cfeec19d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ast\n",
    "import numpy as np\n",
    "tid = \"GRCh38_ENST00000495576\" #for debugging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a5871e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "range(15, 9, -1)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.range(15, 10 - 1, -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5a6069b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_poly_x_regions(sequence, exon_regions = None,  strand = None, nucleotides = 'A', allow_mismatches_without_breaking_streak = 1, return_genomic_coordinates = False):\n",
    "    \"\"\"\n",
    "    Calculate the regions of consecutive nucleotides in a given sequence.\n",
    "\n",
    "    Parameters:\n",
    "    sequence (str): The input nucleotide sequence.\n",
    "    nucleotides (str): The nucleotide(s) to look for. Default is 'A'.\n",
    "    allow_mismatches_without_breaking_streak (int): Number of allowed mismatches within a streak. Default is 1.\n",
    "\n",
    "    Returns:\n",
    "    list of tuples: Each tuple contains the start and end indices of a region.\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "    regions = []\n",
    "    regions_genomic = []\n",
    "    streak_lengths = []\n",
    "    start = None\n",
    "    mismatch_count = 0\n",
    "    streak_length = 0\n",
    "\n",
    "    if return_genomic_coordinates:\n",
    "        if exon_regions is None or strand is None:\n",
    "            raise ValueError(\"exon_regions and strand must be provided when return_genomic_coordinates is True\")\n",
    "        genomic_index = []\n",
    "        for exon_region in exon_regions:\n",
    "            if strand == '+':\n",
    "                genomic_index.extend(np.arange(exon_region[0], exon_region[1] + 1))\n",
    "            else:\n",
    "                genomic_index.extend(np.arange(exon_region[1], exon_region[0] - 1, -1))\n",
    "\n",
    "\n",
    "    for i, nucleotide in enumerate(sequence):\n",
    "        \n",
    "        if nucleotide in nucleotides:\n",
    "            if start is None:\n",
    "                start = i\n",
    "            mismatch_count = 0\n",
    "            streak_length += 1\n",
    "        else:\n",
    "            if start is not None:\n",
    "                mismatch_count += 1\n",
    "                if mismatch_count > allow_mismatches_without_breaking_streak:\n",
    "                    regions.append((start, i - mismatch_count))\n",
    "                    if return_genomic_coordinates:\n",
    "                        regions_genomic.append((int(genomic_index[start]), int(genomic_index[i - mismatch_count]))) # convert to genomic coordinates\n",
    "                    streak_lengths.append(streak_length)\n",
    "                    start = None\n",
    "                    mismatch_count = 0\n",
    "                    streak_length = 0\n",
    "\n",
    "    if start is not None:\n",
    "        regions.append((start, len(sequence) - 1))\n",
    "        if return_genomic_coordinates:\n",
    "            regions_genomic.append((int(genomic_index[start]), int(genomic_index[len(sequence) - 1]))) # convert to genomic coordinates\n",
    "        streak_lengths.append(streak_length)\n",
    "        \n",
    "    if return_genomic_coordinates:\n",
    "        return regions, regions_genomic, streak_lengths\n",
    "    else:\n",
    "        return regions, streak_lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "66f4a611",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_intron_exon_sequences_path = (\"/tmp/Mazutislab-out/Ignas/RT_comparison/genome_and_annotations/lncs_with_single_isoform_fasta_sequences_strand_exon_intron_positions.txt\")\n",
    "transcript_introns = {}\n",
    "transcript_exons = {}\n",
    "transcript_sequences = {}\n",
    "transcript_strands = {}\n",
    "for line in open(transcript_intron_exon_sequences_path):\n",
    "    if line.startswith(\">\"):\n",
    "        transcript_id = line[1:].strip()\n",
    "    elif line.startswith('Strand: '):\n",
    "        strand = line.strip().split(\": \")[1]\n",
    "        transcript_strands[transcript_id] = strand\n",
    "    elif line.startswith('Exon positions:'):\n",
    "        exon_positions = ast.literal_eval(line.strip().split(\": \")[1])\n",
    "        transcript_exons[transcript_id] = exon_positions\n",
    "    elif line.startswith('Intron positions:'):\n",
    "        intron_positions = ast.literal_eval(line.strip().split(\": \")[1])\n",
    "        transcript_introns[transcript_id] = intron_positions\n",
    "    elif line.startswith('Sequence:'): \n",
    "        sequence = line.strip().split(\": \")[1]\n",
    "        transcript_sequences[transcript_id] = sequence\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5a4f8a22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-\n"
     ]
    }
   ],
   "source": [
    "print(transcript_strands[tid])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ca7cfae7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCAGCCTCCCAAGTAGCTGGGGCTACAGGCACCTGCCACCAAACCCGGCTAATTTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTAGCCAGGATCGTCTTGATCTCCTGACCTTGTGATCCACCCGCCTCGGCCTCCCAAATTGCTGGGATTACAGATGTGAGCCACCGCACCTGGTCCAAGAACCCAAGTTTTAGATCTAGAGTGATGTCAGCATGACATTGATTTCCTGAGGCCCAGGGGCGAAGGAGCTGAGGACAGCAGAGGGGTGAAGGAACTCAGCTACAGACAGCAGCAGCTGATGCACAGGCCTCCCAGCGCCTGAAGTCACCCGGAATTGGGAAGTGCTCAGAAGCTTACAAAGCTGCCTCGAGGTGGGAACATGACATAAATCCAAGAGCAGATCCCTGATCCTATAAAAATGTACTAGATGCAGTGGGGGCATTTTAAATGAGCAGAGAAGGACAGACAGATAAACAGAAGGACAAACAGTATTGGGATTGGGATAAATGCTCAGCTTTTGCCCAAATCTTAGTGACTTAAGCATCACTTATTTGCTCACGATTCTGTGGCTGGACCATTTGGTTTGGCTCACAGGGCAGGGACTGTGCTGGTCTTACCTGAGCAGACCTGCATGTCTGCGGTCAACTGGGTTGGCAGAGACAGAGTGACTGTCTTCCTCCAGGAAGCAGCAGGTTAACTGGTTGGCAGAGACAGAGGGACAGAGGGACTGTCTTCCTCCAGGAAGCAGCAGGTTAACTGGTTGGCAGAGACAGAGGGACAGAGGGACTGTCTTCCTCCAGGAAGCAGCAGGTTGGCTCTGTTTCCTTCGTGGGGCAGCTGGTCTCCAGGGCAGCAAGAGAGACCAAGCCCCAGTGCACATTCTACAGCCTCTGTGCACATCAGACTTGTTAATATCCCATTGGCCAGTGTAAGTCACTTGGCCAAGCCCAGATTAAGGAGTGGAAAGATGGAGGCTATCTCCTCCTGGGAGAGGAGGCAAAGGAGGTGGGAGTATTATGTGGCCACTTATGTTTGCAATCTACCATACTTAGCACTTTGAGAAAAGAATTAACTGAGAAACTTGCTTCAAATAGGGCCAGTAAAATGAAGCCCCAATTGAAGTAAAATGCATATATAAAAAATGAAACTGTGACCGATTTTAAGGACAGTATTGGCAAATATTTCTGTGCTCTTGGAGGAGAAGACCCTTATTGGCATGACATGTCAGAAACCACAATGAAAGAATTATTTTAACTTGCATTCATAAAAATTAAAATTATTCATTAAAAACATCGTGAATGAAATTAAAA\n"
     ]
    }
   ],
   "source": [
    "print(transcript_sequences[tid])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2e17daae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, for each transcript, calculate poly-A regions\n",
    "transcript_poly_A_regions = {}\n",
    "for transcript_id, sequence in transcript_sequences.items():\n",
    "    #if transcript_id == tid:#for debugging\n",
    "    if 1==1:\n",
    "        exon_regions = transcript_exons[transcript_id]\n",
    "        strand = transcript_strands[transcript_id]\n",
    "        poly_A_regions, regions_genomic, streak_lengths = calculate_poly_x_regions(sequence, exon_regions=exon_regions, strand=strand, nucleotides='A', allow_mismatches_without_breaking_streak=1, return_genomic_coordinates=True)\n",
    "        #Now for each poly A region, adjust coordinates to genomic coordinates\n",
    "        transcript_poly_A_regions[transcript_id] = {'regions': poly_A_regions, 'regions_genomic': regions_genomic, 'streak_lengths': streak_lengths}\n",
    "#write output to file\n",
    "output_path = (\"/tmp/Mazutislab-out/Ignas/RT_comparison/genome_and_annotations/lncs_with_single_isoform_poly_A_streaks.txt\")\n",
    "with open(output_path, 'w') as f:\n",
    "    for transcript_id, data in transcript_poly_A_regions.items():\n",
    "        f.write(f\">{transcript_id}\\n\")\n",
    "        f.write(f\"Poly-A regions (transcript coordinates): {data['regions']}\\n\")\n",
    "        f.write(f\"Poly-A regions (genomic coordinates): {data['regions_genomic']}\\n\")\n",
    "        f.write(f\"Streak lengths: {data['streak_lengths']}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (star-env)",
   "language": "python",
   "name": "star-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
